lib/blockchain/ethereum.go

```diff
@@ -1220,6 +1277,125 @@ func (e *EthereumClient) AvgBlockTime(ctx context.Context) (time.Duration, error
 	return averageBlockTime, nil
 }
 
+// InitializeHeaderSubscription initializes either subscription-based or polling-based header processing
+func (e *EthereumClient) InitializeHeaderSubscription() error {
+	if e.Client.Client().SupportsSubscriptions() {
+		return e.subscribeToNewHeaders()
+	}
+	// Fallback to polling if subscriptions are not supported
+	e.l.Info().Str("Network", e.NetworkConfig.Name).Msg("Subscriptions not supported. Using polling for new headers.")
+
+	chainPollingRegistry.Lock()
+	poller, exists := chainPollingRegistry.m[e.GetChainID().Int64()]
+	if !exists {
+		poller = &SharedHeaderPoller{
+			chainID:      e.GetChainID().Int64(),
+			pollInterval: 15 * time.Second,
+			done:         make(chan struct{}),
+		}
+		chainPollingRegistry.m[e.GetChainID().Int64()] = poller
+	}
+	e.sharedPoller = poller
+	chainPollingRegistry.Unlock()
+	return e.startHeaderPolling()
+}
+
+// startPollingHeaders starts a polling loop to fetch new headers at regular intervals
+func (e *EthereumClient) startHeaderPolling() error {
+	e.l.Info().Int64("Chain Id", e.GetChainID().Int64()).Msg("Starting Polling Headers")
+	// pollInterval := time.Second * 30
+	chainPollingRegistry.Lock()
+	pollInterval := chainPollingRegistry.m[e.GetChainID().Int64()].pollInterval
+	ticker := time.NewTicker(pollInterval)
+	chainPollingRegistry.Unlock()
```

Comment:
why do we need a lock here? to protect from concurrent read/writes? if so, then maybe you could use RWMutex? and here just use read lock, and where you wrote to `chainPollingRegistry` use a write lock?